### PR TITLE
[#2836] Refactor KafkaProducerFactory

### LINKED ISSUE #2836: KafkaProducerFactory#sharedProducerFactory potentially misleading in how producers are shared
Consider a scenario where you want to use different `KafkaApplicationClient` instances, each working with Kafka configuration properties containing different login credentials, giving access to a specific set of topics.

For a these clients you need a `KafkaProducerFactory` instance. The intuitive way to create one is to use `KafkaProducerFactory#sharedProducerFactory`.

Even when using different `KafkaProducerFactory` instances created with that method for these client, the created Kafka producers with the same name still are getting shared between all client instances. This results in authentication/authorisation errors when working with the producers.

This is especially a problem concerning the command producer created in the `KafkaApplicationClient` because it gets created with a fixed name ("command-sender").

Caching of Kafka producer instances for such a scenario should only be done per `KafkaProducerFactory` instance.

EDIT: made it clear that this is about the `KafkaProducerFactory#sharedProducerFactory` factory method.

### PR DESCRIPTION
**Pull request by b-abel**
This fixes #2836.

### CODE REVIEW

**Comment by b-abel** [comment]
@calohmn Could you have a look. This is the change requested by you. And it requested a lot of reviewers :laughing:

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java [hunk 1]
```diff
@@ -54,17 +54,76 @@
...
+     *            name and the producer configuration.
      */
-    public CachingKafkaProducerFactory(
+    private CachingKafkaProducerFactory(
             final BiFunction<String, Map<String, String>, KafkaProducer<K, V>> producerInstanceSupplier) {
         this.producerInstanceSupplier = producerInstanceSupplier;
     }
 
+    /**
+     * Creates a new factory that produces {@link KafkaProducer#createShared(Vertx, String, Map) sharedFactory
```
Comment (calohmn) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java:68]: "that produces *shared* producers"?

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java [hunk 2]
```diff
@@ -54,17 +54,76 @@
...
+     * producers}. Shared producers can safely be sharedFactory between verticle instances and improve efficiency by
+     * leveraging the batching capabilities of the Kafka client.
+     *
+     * Producers with the same name will be shared (per Vert.x instance), meaning they are shared between instances of
+     * the factory.
+     * <p>
+     * Config must always be the same for the same key in
+     * {@link #getOrCreateProducer(String, KafkaProducerConfigProperties)}.
+     * <p>
```
Comment (calohmn) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java:77]: How about:
```
Instances returned here must use the same config in 
{@link #getOrCreateProducer(String, KafkaProducerConfigProperties)}
invocations that use the same key.
```
?

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java [hunk 3]
```diff
@@ -54,17 +54,76 @@
...
      */
-    public CachingKafkaProducerFactory(
+    private CachingKafkaProducerFactory(
             final BiFunction<String, Map<String, String>, KafkaProducer<K, V>> producerInstanceSupplier) {
         this.producerInstanceSupplier = producerInstanceSupplier;
     }
 
+    /**
+     * Creates a new factory that produces {@link KafkaProducer#createShared(Vertx, String, Map) sharedFactory
+     * producers}. Shared producers can safely be sharedFactory between verticle instances and improve efficiency by
```
[the remark this answers is not in the document]
Comment (b-abel) [reply clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java:69]: Uh, the renaming support of IntelliJ can be a bit aggressive sometimes :smile:

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java (hunk 2 above)
Comment (b-abel) [reply clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/CachingKafkaProducerFactory.java:77]: I think the whole sentence should be removed. This is about the `getOrCreateProducer` method where the JavaDoc already states "The given config is ignored when an existing producer is returned". It always applies to this method regardless of the producer instance supplier.

**Review by calohmn — APPROVED** [review]
LGTM

**Pull request merged**